C++ implementation of temporal memory

rcrowder opened the pull request:
Fixes #314

chetan51 commented on src/nupic/algorithms/Connections.hpp:
```diff
@@ -41,15 +41,15 @@ namespace nupic
-      typedef UInt32 CellIdx;
-      typedef unsigned char SegmentIdx;
-      typedef unsigned char SynapseIdx;
+      typedef UInt CellIdx;
```
Could you explain the motivation behind these type changes?

rcrowder replied:
With the default TM initializes, 65536 cells are required. But that throws the assert check against CELL_MAX (65536-1). The key change here will only be to CELL_MAX, Segment and Synapse indices can remain with the 8 bit storage. I'll change back all but CELL_MAX, and determine a appropriate value for it.

rcrowder replied:
The change to a short SegmentIdx is used in creating an invalid Segment (e.g. in bestMatchingCell and bestMatchingSegment). An alternative way could be implemented.

chetan51 replied:
Okay, the CellIdx change makes sense. Regarding the SegmentIdx change, do we need to create an invalid Segment? Could we just use pointers instead of references in bestMatchingCell / bestMatchingSegment and return null?

rcrowder replied:
Yep, returning a pointer works.

chetan51 commented on src/nupic/algorithms/Connections.hpp:
```diff
@@ -41,15 +41,15 @@ namespace nupic
-      #define CELL_MAX (USHRT_MAX-1)
-      #define SEGMENT_MAX (UCHAR_MAX-1)
-      #define SYNAPSE_MAX (UCHAR_MAX-1)
+      #define CELL_MAX ((USHRT_MAX*32)-1)
```
Can we use UINT_MAX instead?

rcrowder replied:
I've had STL difficulty allocating vector<CellData> that high in the constructor unit test. I'm trying to work out why at the moment.

rcrowder commented:
@chetan51 This looks ready for Capn'Proto and SWIG

rcrowder commented:
@chetan51 cc @scottpurdy @rhyolight I've added the C++ side of the serialization. 

The C++ TM unit test has had the load & save testing depreciated (old method missing on the Random and Connections side) and replaced it with a partial port of the Python testWrite unit test (partial port due to lack of pattern and sequence generators).

I modified the Python algorithms bindings to include a TM interface. Building NuPIC from source, with local core library in Ubuntu 14.04 VM, produces the following Gist output. It shows the output from the three Python TM integration tests (performance, tutorial, extensive, resp.). As far as I'm aware it's using the C++ version with these tests - https://gist.github.com/rcrowder/50b5eddb4a55d98b6f22

Apart from a Doxygen pass, and confirming the integration test results, I can't think what else is required before final review?

rcrowder commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1385 @@
+{
+  version_ = 1;
+
+  // Setup defaults (not all get serialized)
```
This comment is wrong. And with the defaults, was added before realizing that I'd missed out the learningRadius

chetan51 commented:
@rcrowder Regarding the test failures, can you add properties to the Python bindings to `predictiveCells`, `activeCells`, etc. (whatever else `MonitoredTemporalMemory` uses) that returns data from the C++ side? Then those tests should pass.

rcrowder commented:
@chetan51 I've got past that issue by renaming the relevant C++ TM class member variables (dropping underscores). I then faced an issue calling the `TemporalMemoryMonitorMixin` compute function. Changing that to `super(TemporalMemoryMonitorMixin, self).compute(activeColumns, True)` allowed the extensive and tutorial integration tests to call into the C++ (performance test has always ran ok). 
Right now I'm faced with a segmentation fault in the C++ TM computeFn, so will look again at my partial port of the integration tests, and start single step debugging them via Visual Studio to track down the fault.

rcrowder commented:
@chetan51 I've tracked the segmentation fault to the activeColumns sent into C++ TM compute function. Any access to that array in the C++ produces the fault. My temporary Python SWIG code looks similar to the  SP binding, and seems like it should work. I'm guessing that this maybe something NumPY related? 
Right now I'm blocked due to lack of SWIG and NumPY implementation and interaction knowledge. Looks like it may take a while for me to work out how to debug and diagnose SWIG related issues.. I'm currently trying to see whether the NumPY array functions are required to manipulate the array before sending it to the C++ (?).
@rhyolight Could you add the blocked or work-in-progress label to this PR pls?

rcrowder commented:
4 days later and I can get data into the TM with no seg faults. But it looks like getting data out seg faults. I guess with the silence, this is something to discuss further at the hackathon?

rcrowder commented:
@rhyolight 
A progress update before the meeting;

The two NuPIC core library branches [1] and [2] are fairly stable, with only minor recent changes to cope with SWIG bindings. The latter branch containing the monitor mixin and integration tests (a partially complete port, but enough to feed and test TMs using Visual Studio).

1 https://github.com/rcrowder/nupic.core/tree/314-Port-Of-TM
2 https://github.com/rcrowder/nupic.core/tree/314-Port-python-TM

learningRadius has been removed, but I've not looked deeply at the orphan decay changes yet.

Checkout https://github.com/numenta/nupic/issues/2257 for an update on the Python side of this work, including partially complete integration test output.

rhyolight commented:
@rcrowder Very nice, thanks for the report.

rcrowder commented:
@chetan51 @scottpurdy CC @subutai 

Over the last week or so I've made some encouraging steps forward. It's been slow going with the Python TM running through PyCharm debugger in a Ubuntu VM, alongside the C++ TM running via Visual Studio debugger. 

Returning to the use of sets to store cells and column indices has naturally helped, but it also highlighted some issues relating to TM data retention. I resurrected the partial integration tests, and have been stepping through the tutorial tests. This led to the discovery of some algorithmic/Python interpretation issues, mainly Python-specific code and how I translated this to C++. Majority of recent changes have revolved around segment and synapse management, and understanding the differing Connections class implementation.

I've update the tutorial and extensive integration tests output over on Gists. The tutorial test is now down from 3 failures to 1 :) But the extensive test is failing 10 out of 15 :( 
Tutorial - https://gist.github.com/rcrowder/4f7fb87da7c064c35bcf
Extensive - https://gist.github.com/rcrowder/613c597af8527673dda7

Some more defects to find, but seems like only a few hurdles to jump, unless handover is desired sooner?

PS: Could someone update the labels on this PR? The SWIG blockage has been resolved, and the interface seems to be behaving.

rhyolight commented:
Thanks @rcrowder I updated the labels.

rcrowder commented:
@chetan51 cc @rhyolight Looks like two remaining bugs have been squashed! This branch, built locally, and used by NuPIC in a Ubuntu VM, has all the integration tests now passing!!

I modified the abstract class to output the TM CPP parameters, and have updated the three C++ Gists
* Performance - https://gist.github.com/rcrowder/9e3142fc78fd393430c0
* Tutorial - https://gist.github.com/rcrowder/4f7fb87da7c064c35bcf
* Extensive - https://gist.github.com/rcrowder/613c597af8527673dda7

I need to go over Doxygen tags, but other than that I think this is ready for a review, phew!

chetan51 commented:
That's great @rcrowder! Awesome.

Before I review it, can you merge with master? There were some changes in Connections that we want to make sure play well with your changes here. Thanks!

rcrowder commented:
@chetan51 Definitely, I'll do that next. There is also a few other slight differences that I've thought off that can be ironed out too.

rcrowder commented on appveyor.yml:
```diff
@@ -128,7 +128,7 @@ after_build:
 
-      7z a -ttar -y -bd nupic_core-$env:APPVEYOR_REPO_COMMIT-windows64.tar .
+      7z a -ttar -y -bd nupic_core-$env:APPVEYOR_REPO_COMMIT-windows64.tar . | findstr /b /c:"Everything is Ok" /c:"Scanning" /c:"Creating archive"
 
```
This reduces the CI build log output considerably, all the "Compressing ..." lines.

chetan51 replied:
@rhyolight Please review this line.

rcrowder commented on src/CMakeLists.txt:
```diff
@@ -541,6 +543,7 @@ set(TEST_HEADERS
     test/unit/algorithms/NearestNeighborUnitTest.hpp
     test/unit/algorithms/SpatialPoolerTest.hpp
+    test/unit/algorithms/TemporalMemoryTest.hpp
     test/unit/engine/InputTest.hpp
```
Python TM unit test has been left pointing to the Python, and this is the port to check the C++ side.

rcrowder commented:
@chetan51 Merge complete. Two main conflicts were with TM proto differences, and the new Connections consts. 
I've also made a pass over the Doxygen syntax, and ran it through a localhost server to make sure the html looks ok. Also see recent commits to the nupic TM-Bindings branch and PR. 
I can only see a few 80 chars per line convention breaks, but could leave those until a potential future review commit.

chetan51 commented on src/nupic/algorithms/Connections.cpp:
```diff
@@ -59,7 +59,7 @@ Segment Connections::createSegment(const Cell& cell)
   Segment segment(segments.size(), cell);
 
-  if (segments.size() == maxSegmentsPerCell_)
+  if (segments.size() == (size_t)maxSegmentsPerCell_)
```
Would it be better to use `(SegmentIdx)` here instead?

rcrowder replied:
Changed

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+
+tuple<vector<Segment>, set<Cell>> TemporalMemory::computePredictiveCells(
+  set<Cell>& _activeCells,
+  Connections& _connections)
```
Why doesn't this just use `computeActivity`? See `fast_temporal_memory.py`.

rcrowder replied:
I have this change locally. Will compare output logs in the morning, and try the other change below. All tests are still passing, but there's obvious divergence in synaptic creation in the tutorial tests, and subsequent extensive test statistics.

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+// ==============================
+
+tuple<Cell*, Segment*>
+TemporalMemory::bestMatchingCell(
```
This could use `Connections.mostActiveSegmentForCells` instead of reimplementing it.

chetan51 commented on src/nupic/algorithms/Connections.hpp:
```diff
@@ -332,6 +332,15 @@ namespace nupic
+         *
+         * @return (set)Synapse indices
+         */
+        std::vector<Synapse> synapsesForPresynapticCell(const Cell& presynapticCell) const;
```
This is not necessary; see comment https://github.com/numenta/nupic.core/pull/409/files#r34511536.

rcrowder commented:
@chetan51 Thanks for starting a review. Looks like a minor cast change and optional work so far, which is encouraging. As you can see the Python TM code was frozen at a particular point. With further work to cope with the fast TM Python changes, and a few other changes since then (Orphan decay, etc.).
Since starting the port my life has had some big changes, I'm running out of time and money to contribute short term. So imo I'd be better to get this working port merged in, breaking the link with myself, and allowing anyone subsequently to add the optional items and missing recent changes.
I'm hoping over the next few days we can get this approved and merged.

chetan51 commented on src/nupic/algorithms/TemporalMemory.hpp:
```diff
@@ -0,0 +1,613 @@
+        */
+        TemporalMemory(
+          vector<UInt> columnDimensions,
+          Int cellsPerColumn = 32,
```
Why are these and other `Int`s below not `UInt`s?

rcrowder replied:
Except seed, all updated to UInt

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+  if (cellsPerColumn <= 0)
+    NTA_THROW << "Number of cells per column must be greater than 0";
+
+  // TODO: Validate all parameters (and add validation tests)
```
Please create a PR for this and reference it here

chetan51 replied:
bump

rcrowder replied:
https://github.com/numenta/nupic.core/issues/504 and a comment about it placed in code below this comment.

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+  winnerCells = _winnerCells;
+  activeSegments = _activeSegments;
+  predictiveCells.clear();
+  for (Cell c : _predictiveCells)
```
Why does `predictiveCells` do something different from the other sets?

rcrowder replied:
predictiveCells is the only public class member variable that the Python expects to be iterable. Having it as a vector seemed to be the best way to do that through SWIG.

chetan51 replied:
I thought all these variables are iterable:

    self.activeCells = set()
    self.predictiveCells = set()
    self.activeSegments = set()
    self.winnerCells = set()

rcrowder replied:
I had these all as sets originally but had issues with SWIG. Switching to a vector helped, but I can go back and work out why SWIG and sets are different from vectors?

chetan51 replied:
Are all those variables iterable in Python through SWIG? If so, no worries, we don't have to change it now.

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+  Connections& _connections,
+  bool learn)
+{
+  set<UInt> _activeColumns;
```
Can we just use the ones that are passed in? Why do we need to copy them?

rcrowder replied:
It was a while since I looked at this, but recall it relates to in-memory changes to those given sets and helped out with the copied data. This, and other data manipulations, are rife for optimization.

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+      permanence -= permanenceDecrement_;
+
+    // Keep permanence within min / max bounds
+    //permanence = max(0.0, min(1.0, permanence));
```
Remove commented-out code

rcrowder replied:
Done

chetan51 commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -0,0 +1,1093 @@
+
+Int TemporalMemory::numberOfColumns(void)
+{
+  Int acc(1);
```
Don't we have a `numColumns_` variable?

rcrowder replied:
Yes, but I tried to stick to the Py as close as possible to easy review. I've changed this to return that member variable, and inlined the function proto.

chetan51 commented on src/nupic/types/Types.h:
```diff
@@ -36,9 +36,10 @@
 #pragma warning( disable : 4251 ) // needs to have dll-interface to be used by clients of class 
 #pragma warning( disable : 4275 ) // non dll-interface struct used as base for dll-interface class
+#pragma warning( disable : 4305 ) // truncation from 'double' to 'nta::Real', possible loss of data (LOTS of various type combinations)
+#pragma warning( once : 4838 ) // narrowing conversions
```
@scottpurdy Please review this change.

scottpurdy replied:
how is 4305 different from 4244 above? And I don't understand 4838 well enough - where does it occur in the code?

rcrowder replied:
4305 and 4244 covers very similar things, similar to 4838. These pagmas have mainly been used to reduce the build output. Any numbers destined for a float container, but don't have the 'f' suffix is treated as a double, and then can try and suggest multiple STL riddled pages of what to try.. I put 4838 as a once report (per file), to make sure that not all the warnings are suppressed.

chetan51 replied:
@scottpurdy Is this change okay then?

scottpurdy replied:
I don't know these different warnings well enough so fine either way.

chetan51 commented on src/test/unit/algorithms/TemporalMemoryTest.cpp:
```diff
@@ -0,0 +1,943 @@
+
+    // Depreciated serialization method (incomplete due to lack of load & save 
+    // in Connections and Random classes). Replaced by Cap'n Proto read & write.
+    //testSaveLoad();
```
@scottpurdy Should we remove this test code and the deprecated function if they're not going to be used anyway?

scottpurdy replied:
save and load are the only way to serialize right now so they need to work until we make the switch to capnp

chetan51 replied:
Then should this test be enabled and passing?

scottpurdy replied:
Yes, although that is totally fine as a follow up PR to speed this one up.

rcrowder replied:
testSaveLoad can be enabled (commented out lines in load and save), but requires Connections and Random to support the old serialization method. Could someone create separate PRs for those two changes?

chetan51 commented on src/test/unit/algorithms/TemporalMemoryTest.cpp:
```diff
@@ -0,0 +1,943 @@
+    TemporalMemory tm1, tm2;
+
+    tm1.initialize({ 100 }, 4, 7, 0.37, 0.58, 4, 18, 0.23, 0.08, 91);
+    /*
```
Please file a separate issue to finish the rest of this test, and reference it here.

chetan51 replied:
bump

rcrowder replied:
Added https://github.com/numenta/nupic.core/issues/505 and comment in the code.

chetan51 commented:
Finished with initial review.

rcrowder commented:
Thank you for the prompt review. I've addressed most of the issues in the last commit. Not sure about the Travis failure (AppVeyor passes!). I'm back to stepping through the Py TM tutorial test and the C++ partial, but functional, port of that test. Hunting down the difference in synaptic creation.

rcrowder commented:
See https://github.com/numenta/nupic.core/issues/419 for an update on the state of the integration tests and comparison to the Python output.

chetan51 commented:
@rcrowder Merged! Thank you for your work on this.

rcrowder commented:
@chetan51 Superb! Thanks for swiftly jumping on this. As you may have seen I've added a PR with changes for the synapse decay, and a NuPIC PR with my SWIG bindings (that needs some assistance to complete). I'll continue looking at the divergence in Extensive test stats though.

rhyolight commented:
:fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks:
